### DISCUSSION TITLE
How can we set text content (`textContent` (`innerHTML`)) of fragments?

### DISCUSSION THREAD

**Discussion by kireerik** [solidjs/solid#1759] [General]
related documentation:
- https://www.solidjs.com/docs/latest/api#innerhtmltextcontent

**Comment by AlexErrant** [comment]
Not sure I follow - fragments aren't "real" HTML elements, and so can't have `textContent` or `innerHtml`. They're just JSX helpers - actually really just a list (of elements). In the same way you can't return multiple things from a javascript function (unless its in a list), you can't return multiple elements from a javascript function... unless they're in a list. Fragment is that list.

**Reply by kireerik** [reply]
Yes, I don't need a (the) HTML element that's why I want to use a `fragment` to return other HTML that I set. Seems like for example [qwik](https://qwik.builder.io/) has [`HTMLFragment`](https://qwik.builder.io/api/qwik/#htmlfragment).

**Reply by AlexErrant** [reply]
Not familiar with Quik, and those docs aren't enough for me to realllllyy understand what it does, but like, you can just put a textNode and/or component in a fragment. https://playground.solidjs.com/anonymous/5204aadb-30c0-4901-866d-0d724ab93b4f

**Reply by ryansolid** [reply]
Yeah unlike Qwik we don't have a Virtual DOM. While I went the DocumentFragment abstraction initially it was a lot worse for performance. Solid understands real DOM nodes but it doesn't understand HTML. Our fragments aren't even components. They are arrays.

The easiest way to do this right now is a bit imperative. Obviously text in fragments just works but HTML strings will not. So you can use `template` from `solid-js/web` function. And it will convert text to DOM nodes. This obviously comes with the whole you better trust the string warning as the intention is to get HTML.

You could do something like:
```js
import { template } from "solid-js/web";

function HtmlFragment(props: { children: string }) {
  return <>{template(props.children)()}</>
}
```

use it like:
```js
<HtmlFragment>{`<h1>hello</h1>`}</HtmlFragment>
```
(note the expression and string... otherwise it will see angle brackets and assume it is JSX)

or:
```js
<HtmlFragment>{htmlString()}</HtmlFragment>
```

Again make sure you know what you are doing when using this component.

**Reply by AlexErrant** [reply]
Tagging with `dangerouslySetInnerHTML` to help future searchers.

**Reply by kireerik** [reply]
Yes, I know, which is great. It is nice to know that this is possible. Maybe it is good to add the `template` function to the documentation sometime.

**Reply by kireerik** [reply]
I can just simply do `template('<h1>hello</h1>')`.

**Reply by kireerik** [reply]
I see this is imported from `/web`. Does this works on the server side as well?

**Reply by kireerik** [reply]
On the `server` side this works: `{t: '<h1>hello</h1>'}`

**Comment by snnsnn** [comment]
Fragments are not runtime but compile time constructs, so you UI does not have concept of fragment. Fragments are needed for jsx parsers. 

You can write your components using fragments, but it will introduce overhead because of the extra component layer:

```tsx
import { Component, createComponent, createSignal } from 'solid-js';
import { render } from 'solid-js/web';

const Comp: Component<{ children: number }> = (props) => {
  return <>{props.children}</>;
};

const App = () => {
  const [count, setCount] = createSignal(0);
  setInterval(() => setCount((p) => p + 1), 1000);

  return (
    <div>
      <Comp>{count()}</Comp>
    </div>
  );
};

render(App, document.body);
```

You can get the same result returning a function:

```tsx

const Comp: Component<{ children: number }> = (props) => {
  return createMemo(() => props.children) as unknown as JSXElement;
};
```

This works because Solid implicitly invokes the returned callback.

Alternatively you can set the innerText or innerHTML of an element directly using an effect but that would not work for SSR. Also setting innerHTML directly requires sanitizing because you would expose your app to XSS attacks.

```ts
import { createSignal } from 'solid-js';
import { render } from 'solid-js/web';

const App = () => {
 const [count, setCount] = createSignal(0);
 setInterval(() => setCount((p) => p + 1), 1000);
 return <div innerHTML={count().toString()}></div>; 
};
render(App, document.body);
```
